_posts/2024-06-14-remote-code-execution-engine.md

```diff
@@ -0,0 +1,1426 @@
+    working_dir: /app
+    environment:
+      CLUSTER_IP: 'worker-1'
+      CLUSTER_PORT: '17350'
+      CLUSTER_ROLE: 'worker'
+      SEED_IP: 'master'
+      SEED_PORT: '1600'
+  worker-2:
+    image: braindrill:latest
+    build:
+      context: .
+      dockerfile: Dockerfile
+    tty: true
+    stdin_open: true
+    ports:
+      - '17351:17351'
+    security_opt:
+      - 'seccomp=unconfined'
+    volumes:
+      - /var/run/docker.sock:/var/run/docker.sock
+      - engine:/data
+    working_dir: /app
+    environment:
+      CLUSTER_IP: 'worker-2'
+      CLUSTER_PORT: '17351'
+      CLUSTER_ROLE: 'worker'
+      SEED_IP: 'master'
+      SEED_PORT: '1600'
+  worker-3:
+    image: braindrill:latest
+    build:
+      context: .
+      dockerfile: Dockerfile
+    tty: true
+    stdin_open: true
+    ports:
+      - '17352:17352'
+    security_opt:
+      - 'seccomp=unconfined'
+    volumes:
+      - /var/run/docker.sock:/var/run/docker.sock
+      - engine:/data
+    working_dir: /app
+    environment:
+      CLUSTER_IP: 'worker-3'
+      CLUSTER_PORT: '17352'
+      CLUSTER_ROLE: 'worker'
+      SEED_IP: 'master'
+      SEED_PORT: '1600'
+
+volumes:
+  engine:
+    external: true
+```
+
+This `docker-compose.yaml` file sets up a `Pekko` cluster with one `master` and three `worker` services. 
+
+Key points include:
+
+1. Version: Uses Docker Compose version 3.
+2. Services: Defines four services - `master`, `worker-1`, `worker-2`, and `worker-3`.
+
+#### Master:
+- Image: Uses `braindrill:latest` and builds from the Dockerfile in the current context. `braindrill` image will be built using a special `Dockerfile` definition which will be covered later. For better automation, we could also write a deployment shell script that automates all those details. 
+- Container Name: `master` - in Pekko terms - master node
+- Ports: Maps `1600:1600` and `8080:8080`
+- Volumes: Mounts Docker socket for `DooD` and `engine` volume for sharing temporary files (e.g `python123.py`, `Program123.java`) where user code will be written.
+- Environment Variables: Sets cluster details (`CLUSTER_IP`, `CLUSTER_PORT`, `CLUSTER_ROLE`, `SEED_IP`).
+
+#### Workers:
+- Image: Uses `braindrill:latest` and builds from the Dockerfile in the current context.
+- Ports: Each worker maps a unique port (`17350`, `17351`, `17352`).
+- Security: Disables default security profiles (`seccomp=unconfined`).
+- Volumes: Mounts Docker socket for `DooD` and `engine` volume for sharing temporary files where user code will be written.
+- Environment Variables: Sets cluster and seed details for each worker.
+
+#### Common Settings:
+- TTY and stdin_open: Enabled for interactive use.
+- Working Directory: `/app`.
+
+#### Volumes:
+- Engine: An external volume named `engine` which is necessary to share temporary program files that will be later executed.
+
+### 3.4 Dockerfile
+
+`Dockerfile` is a declarative file for building a `docker` image.
+
+Pretty simple and self-explanatory:
```

Comment:
Avoid this kind of phrasing - something is self-explanatory only when the reader is an expert on the subject, in which case they might not need to read the article 😄